### ISSUE TITLE
Factor out a common supertype for Button Nodes

### ISSUE THREAD

**Issue by zepumph** [phetsims/sun#643]
From many issues, like https://github.com/phetsims/sun/issues/515, it seems to be time to consolidate sun buttons into a central common code Node. This should be independent of view (what it looks like), and just hold logic that applies to all buttons. I don't think the refactor will be that challenging, and will likely help us get a better enabledProperty API over in https://github.com/phetsims/sun/issues/515. 

Tagging @pixelzoom

**Comment by zepumph** [comment]
I have a pretty much successfully working refactor into `ButtonNode`, which RectangularButtonView and RoundButtonView both extend from. 

It houses:
* creating the listener from the buttonModel
* setting the buttonModel
* creating the `baseColorProperty`
* Helper methods for these three features.

I think I will commit soon, because it is working well, and may unblock weirdness over in #515. That said, Here is what I am uncertain about:

* It would be nice to factor out `content` too, as there is some duplication there. 
* ButtonNode doesn't mutate options, but instead it relies on subtypes to do it. This is because the sub-type views rely on calling mutate after the content has been set, as some formatting/layout relies on having content first. Even so, we may be able to refactor the subclasses to better support this, like perhaps passing up children or something. 

all three local aqua fuzzings are running well. Commit coming soon.

**Comment by zepumph** [comment]
Blocking publication until CT has its way with this commit, and this can be discussed/reviewed.

**Comment by pixelzoom** [comment]
Should this base class be called `ButtonView`, in keeping with the `RectangularButtonView and `RoundButtonView` names? Or call it`ButtonNode`, and rename the subclasses to `RectangularButtonNode` and `RoundButtonNode`?  I think I prefer the latter, I've never liked the "View" suffix here.

**Comment by zepumph** [comment]
Agreed.
- [ ] Rename RectangularButtonView -> Node
- [ ] Rename RoundButtonView -> Node

**Comment by pixelzoom** [comment]
I'm pretty sure that this should be removed from RoundButtonView, since it's handled by EnabledNode:

```js
    // PDOM - indicate to screen readers that the button is not clickable
    const updatePDOMEnabled = enabled => {
      this.setAccessibleAttribute( 'aria-disabled', !enabled );
    };
    buttonModel.enabledProperty.link( updatePDOMEnabled );
```

**Comment by pixelzoom** [comment]
RoundButtonModel and RectangularButtonModel are now setting options twice in constructors:

```
super( buttonModel, options );
...
this.mutate( options );
```

This will really mess things up for transform-related options, and perhaps for other options.

**Comment by pixelzoom** [comment]
- [ ] Move handling of pickable to ButtonNode

RoundButtonView does this, which seems redundant:

```js
    if ( content ) {
      content.pickable = false;
    }

    // Set pickable such that sub-nodes are pruned from hit testing.
    this.pickable = null;
```

RectangularButtonView does this, which seems wrong, because the content is actually `alignBox`:

```js
    let alignBox = null;
    if ( content ) {

      // For performance reasons, the content should be unpickable.
      content.pickable = false;

      // align content in the button, this AlignBox must be disposed since it adds listener to content bounds
      alignBox = new AlignBox( content, {
        alignBounds: new Bounds2(
          options.xMargin,
          options.yMargin,
          button.width - options.xMargin,
          buttonHeight - options.yMargin
        ),
        xAlign: options.xAlign,
        yAlign: options.yAlign
      } );
      this.addChild( alignBox );
    }
```

We should probably remove all of this and just do this in ButtonNode:

```js
    // Set pickable such that sub-nodes are pruned from hit testing.
    this.pickable = null;
```

**Comment by pixelzoom** [comment]
- [ ] 

It's difficult to see what else RectangularButtonView and RoundButtonView might have in common, because their constructors do things in different orders.  I'm going to clean that up.

**Comment by pixelzoom** [comment]
- [ ] Move `interactionStateProperty` listener to ButtonNode

After cleaning up constructors, I see this code duplicated in RectangularButtonView and RoundButtonView. Can it be moved to ButtonNode?

```js
    // Control the pointer state based on the interaction state.
    const handleInteractionStateChanged = state => {
      this.cursor = state === ButtonInteractionState.DISABLED ||
                    state === ButtonInteractionState.DISABLED_PRESSED ? null : 'pointer';
    };
    interactionStateProperty.link( handleInteractionStateChanged );
```

**Comment by pixelzoom** [comment]
- [ ] ButtonNode `addListener` and `removeListener` will fail. Delete them? 

ButtonModel has these methods:

```js
  // @public
  addListener( listener ) {
    this.buttonModel.addListener( listener );
  }

  // @public
  removeListener( listener ) {
    this.buttonModel.removeListener( listener );
  }
```

But ButtonModel has no `addListener` and `removeListener` methods.  I verified that this bug was not introduced in https://github.com/phetsims/sun/commit/b606357bca9ae477f13f30e7a2264f5b7b674303. I I suspect that we can just delete these methods.

**Comment by pixelzoom** [comment]
- [ ] `baseColor` option belongs in ButtonNode

It's currently duplicated in the options for RectangularButtonView and RoundButtonView.  And there's a bug in both RectangularButtonView and RoundButtonView where `this.baseColorProperty` should be used to create the button's background Node.

**Comment by pixelzoom** [comment]
- [ ] EnabledNode and appearanceStrategy are both changing the button's appearance when disabled.

This is causing the button to look doubly disabled.  We should delete `options.disabledBaseColor` and remove this responsibility from appearanceStrategy.

**Comment by pixelzoom** [comment]
- [ ] Change cursor by listening to `enabledProperty`:

Above I moved the duplicate interactionStateProperty listener into ButtonNode:

```js
    // Control the pointer state based on the interaction state.
    const interactionStateListener = state => {
      this.cursor = state === ButtonInteractionState.DISABLED ||
                    state === ButtonInteractionState.DISABLED_PRESSED ? null : 'pointer';
    };
    interactionStateProperty.link( interactionStateListener ); // unlink in dispose
```

I also suspect that this code is totally unnecessary.  When `enabledProperty` is false, it's not pickable, and the cursor is not shown.  This is also also buggy, because it sets cursor to 'pointer', effectively blowing away any cursor that might have been set via options.

If this is for some reason necessary, it should be changed to a listener for `this.enabledProperty`. And it should save/restore the cursor that was specified via options.

**Comment by pixelzoom** [comment]
Re renaming `RectangularButtonView` and `RoundButtonView`....  Here are some of their subclasses:

```
class RectangularPushButton extends RectangularButtonView 
class RectangularToggleButton extends RectangularButtonView
...
class RoundPushButton extends RoundButtonView
class RoundToggleButton extends RoundButtonView
...
```

In this content, `RectangularButton` and `RectangularButton` (with the "Node" suffix) seem like better names. 

With abstract base class `ButtonNode` or `Button` or `SunButton` or ... ?

**Comment by pixelzoom** [comment]
I did a ton of work here.   The only issue that I haven't addressed is https://github.com/phetsims/sun/issues/643#issuecomment-715478451:

> - [ ]  options are set twice

This is causing problems in layout of buttons, because transform options will be applied twice.  A couple of examples (there are likely many others):

In Natural Selection, the environment radio buttons are messed up:

<img width="165" alt="screenshot_593" src="https://user-images.githubusercontent.com/3046552/97063535-873d1080-155d-11eb-882c-dd97ff81b697.png">

In scenery-phet demo _Buttons_ screen, the "Change some colors" button is supposed to be just above the "Message" text:

<img width="1048" alt="screenshot_592" src="https://user-images.githubusercontent.com/3046552/97063524-6bd20580-155d-11eb-8d4e-9238aa511c74.png">

**Comment by zepumph** [comment]
Some review this far: 

- [ ] buttonAppearanceStrategy needs type doc, as the interaface for the option is sorta scattered around. 
- [ ] Can you comment on why contentAppearanceStragety was no longer needed?

I agree that dealing with calling `options` multiple times is the next step.

**Comment by pixelzoom** [comment]
Documentation for `buttonAppearanceStrategy` option added in the above commit. Rather than a complicated type expression, I just referred to an example, e.g.:

```js
      // Class that determines the button's appearance for the values of interactionStateProperty.
      // See RoundButton.ThreeDAppearanceStrategy for an example of the interface required.
      buttonAppearanceStrategy: RoundButton.ThreeDAppearanceStrategy,
```

There's also the possibility that some values of `buttonAppearanceStrategy` won't need to implement `dispose`, so I made that conditional, e.g.:

```js
    // @private
    this.disposeRoundButton = () => {
      buttonAppearanceStrategy.dispose && buttonAppearanceStrategy.dispose();
    };
```

@zepumph look OK to you?

**Comment by samreid** [comment]
There is a failure in CT in Wave on a String. 

```
wave-on-a-string : fuzz : unbuilt
https://bayes.colorado.edu/continuous-testing/ct-snapshots/1603548360340/wave-on-a-string/wave-on-a-string_en.html?continuousTest=%7B%22test%22%3A%5B%22wave-on-a-string%22%2C%22fuzz%22%2C%22unbuilt%22%5D%2C%22snapshotName%22%3A%22snapshot-1603548360340%22%2C%22timestamp%22%3A1603549377497%7D&brand=phet&ea&fuzz&memoryLimit=1000
Query: brand=phet&ea&fuzz&memoryLimit=1000
Uncaught Error: Assertion failed: alignBounds should be a non-empty finite Bounds2
Error: Assertion failed: alignBounds should be a non-empty finite Bounds2
at window.assertions.assertFunction (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1603548360340/assert/js/assert.js:22:13)
at AlignBox.setAlignBounds (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1603548360340/scenery/js/nodes/AlignBox.js:138:15)
at AlignBox.set alignBounds [as alignBounds] (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1603548360340/scenery/js/nodes/AlignBox.js:152:35)
at https://bayes.colorado.edu/continuous-testing/ct-snapshots/1603548360340/scenery/js/nodes/Node.js:5349:23
at arrayEach (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1603548360340/sherpa/lib/lodash-4.17.4.js:537:11)
at Function.forEach (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1603548360340/sherpa/lib/lodash-4.17.4.js:9359:14)
at AlignBox.mutate (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1603548360340/scenery/js/nodes/Node.js:5336:7)
at AlignBox.Node [as constructor] (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1603548360340/scenery/js/nodes/Node.js:581:10)
at new AlignBox (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1603548360340/scenery/js/nodes/AlignBox.js:85:8)
at new WOASScreenView (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1603548360340/wave-on-a-string/js/wave-on-a-string/view/WOASScreenView.js:205:20)
id: Bayes Chrome
Snapshot from 10/24/2020, 8:06:00 AM
```

WOASScreenView has this code:

```js
    const resetAllButton = new ResetAllButton( {
      listener: () => model.reset(),
      right: this.layoutBounds.maxX - 10,
      bottom: this.layoutBounds.maxY - 10,
      tandem: tandem.createTandem( 'resetAllButton' )
    } );
    resetAllButton.scale( 0.924 );
    this.addChild( resetAllButton );
```

Unfortunately, the resetAllButton does not have the right location, even though `this.layoutBounds.maxX` is correct.  This means there is a problem with the positioning.  Looking into RoundButton.js, I can see that the options are passed through to the `super` call before the child `button` is added.  Therefore it cannot be positioned properly.  Typically this is solved by calling mutate after the children have been added.

This could disrupt layouts, so should be marked blocks publication.  I noticed it is already marked as such.

**Comment by zepumph** [comment]
That is correct. Please see https://github.com/phetsims/sun/issues/643#issuecomment-715642747. That is the last large piece of regression that we are sorting out right now.

**Comment by zepumph** [comment]
Above I have been trying to move more logic into ButtonNode. My end-goal hope is that we can mutate options only at the end of ButtonNode, and that it will fix all the layout problems currently caused by calling options multiple times. To get there, we may have to move content up to the super type first, and align options. That is probably the most complete.

Above, I moved up a couple of things:
1. The buttonAppearanceStrategy option and creation
2. I pass up the `buttonBackground` (either a circle or rectangle). This way the supertype can create the appearance strategy, and set the fill/pickable consistently.
3. See https://github.com/phetsims/sun/issues/647 for work factoring out the FlatAppearanceStrategy in both Round/Rectangle (which were identical) into ButtonNode.

Next I will look at content/AlignGroup/ and unifying some options.

**Comment by zepumph** [comment]
- [ ] lineWidth can be factored out to the superType, at which point, we could turn the setters on buttonBackground into a `mutate` call, since there are three options applied consistently to the passed in buttonBackground.

**Comment by zepumph** [comment]
In the above, I was able to factor out `content` wrapping and adding as a child into ButtonNode. There are a lot of TODOs in the code, but none felt so large that they stopped me from committing what I had. Layout is now working, and RoundButton and Rectangular button inject many more dependencies to the super type. @pixelzoom I would love you to take a look at these changes, and TODOs, as I won't be able to work on this again until Wednesday. 

the CT error from Wave Interference is fixed, and layout of buttons is fixed (I saw problems in NS, MAL, sun demo on buttons screen, states of matter and likely more). All the layout I see now is fixed.

**Comment by zepumph** [comment]
The above commit fixed a couple of CT errors that looked like:

```
Uncaught Error: Assertion failed: Cannot add a listener to a disposed TinyEmitter
Error: Assertion failed: Cannot add a listener to a disposed TinyEmitter
at window.assertions.assertFunction (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1603567870839/assert/js/assert.js:22:13)
at TinyEmitter.addListener (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1603567870839/axon/js/TinyEmitter.js:98:15)
at PaintColorProperty.lazyLink (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1603567870839/axon/js/Property.js:401:25)
at PaintObserver.secondaryLazyLinkProperty (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1603567870839/scenery/js/display/PaintObserver.js:229:16)
at PaintObserver.attachPrimary (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1603567870839/scenery/js/display/PaintObserver.js:122:12)
at PaintObserver.attachPrimary (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1603567870839/scenery/js/display/PaintObserver.js:136:14)
at PaintObserver.setPrimary (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1603567870839/scenery/js/display/PaintObserver.js:60:12)
at PathSVGDrawable.proto.markDirtyFill (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1603567870839/scenery/js/display/drawables/PaintableStatefulDrawable.js:121:25)
at Path.invalidateFill (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1603567870839/scenery/js/nodes/Paintable.js:912:30)
at Path.setFill (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1603567870839/scenery/js/nodes/Paintable.js:133:16)
id: Bayes Chrome
Snapshot from 10/24/2020, 11:31:10 AM
```

I think it was where the baseColorPropety and/or the pressListener were disposed before the buttonAppearanceStrategy. I couldn't reproduce locally after the above.

**Comment by pixelzoom** [comment]
@zepumph can this item in https://github.com/phetsims/sun/issues/643#issuecomment-715478451 be checked off? 

> - [ ]  options are set twice

Looks like you have addressed it.  Are there still any layout problems with buttons? Or do we need to confirm that, and (if so) how?

**Comment by zepumph** [comment]
I checked it off just now. There are now only 1 set of options being applied, in ButtonNode. I was able to see that all the buggy usages that were reported and that I saw are fixed in master:

* Reset all button layout where the reset all button was on the top and left (0,0) (in SOM atomic interactions screen and MAL)
* NS environment button that you pointed out in https://github.com/phetsims/sun/issues/643#issuecomment-715642747
* "Change some button colors" button in sun demo that you pointed out in https://github.com/phetsims/sun/issues/643#issuecomment-715642747
* CT failure in Waves interference from https://github.com/phetsims/sun/issues/643#issuecomment-715994650.

It is possible that regressions have been added, but unlikely, as I tested the look and behavior in the sun demo (before/after), scenery phet demo (before/after), and fuzzed all sims. CT also looks clear. Do you have any recommendations about further testing? Also CT is clear. 

Next steps for this issue:

* tackle https://github.com/phetsims/sun/issues/643#issuecomment-716039038
* A review of my changes, I think it would be good to try to unblock this for publication eagerly.

**Comment by zepumph** [comment]
>lineWidth can be factored out to the superType, at which point, we could turn the setters on buttonBackground into a mutate call, since there are three options applied consistently to the passed in buttonBackground.

Because both subtypes use `lineWidth` and `stroke` in an appearance strategy specific to that subtype, I couldn't remove those options from the sub types. Thus, this feels a little moot.

@pixelzoom please take a look at this patch to see if you think we should apply it, or just leave things as they are.


Furthermore all the work I did this weekend is complete and ready for review.

**Comment by jessegreenberg** [comment]
I noticed that opacity of content for deselected RadioButtonGroupMembers is no longer changing and tracked it down to https://github.com/phetsims/sun/commit/4f6ed2827256acd1f10e1bc0c524731db62520b4#diff-847130871035dab4cf7080a2b28b5fe5ec50c7dd3f45e13652e98e4bba0a0963 referencing this issue. Was this change intentional? If so, can the RadioButtonGroupMember options relating to content opacity be removed?

**Comment by jonathanolson** [comment]
It looks like the forward/back buttons now look uncentered (a regression):

![image](https://user-images.githubusercontent.com/480878/97325461-a1663f80-1838-11eb-928c-46e81ac6212a.png)

**Comment by pixelzoom** [comment]
Looks like that change to RadioButtonGroupMember was not intentional. EnabledNode should have responsibility for enabled state, but RadioButtonGroupMember should have responsibility for selected state.

**Comment by pixelzoom** [comment]
> ... @pixelzoom please take a look at this patch to see if you think we should apply it, or just leave things as they are.

This should definitely not be handled by ButtonNode.  buttonBackground is {Node}, which does not have lineWidth.   And for the same reason, this line also does not belong in ButtonNode:

`    buttonBackground.fill = this.baseColorProperty;`

**Comment by pixelzoom** [comment]
> ... @pixelzoom please take a look at this patch to see if you think we should apply it, or just leave things as they are.

lineWidth should definitely not be handled by ButtonNode.  buttonBackground is {Node}, which does not have lineWidth.   And for the same reason, this line also does not belong in ButtonNode:

`    buttonBackground.fill = this.baseColorProperty;`

But I think we can let baseColorProperty slide (for now? forever?), because it's needed by buttonAppearanceStrategy.

**Comment by zepumph** [comment]
>But I think we can let baseColorProperty slide (for now? forever?), because it's needed by buttonAppearanceStrategy.


How about this updated type doc that expects buttonBackground to be able to support fill?

I agree the lineWidth is good where it is.

**Comment by pixelzoom** [comment]
"this updated type doc" is:

```js
@param {Paintable} buttonBackground - the background of the button (like a circle or rectangle).
```

That resolves the issue of usages of the Paintable API.  But it does not cover uses of the Node API, like:

```js
109 buttonBackground.pickable = false;
129 alignBounds: buttonBackground.bounds,
```

**Comment by pixelzoom** [comment]
Over in Slack developer channel, we decided that type expressions involving mixins and traits should use the general form `{Class,Mixin,Trait,...}`.   @zepumph cited https://github.com/jsdoc/jsdoc/issues/1199#issuecomment-492586767.  So in this case:

```js
@param {Node,Paintable} buttonBackground - the background of the button (like a circle or rectangle).
```

**Comment by pixelzoom** [comment]
It looks like any remaining work is being tracked in other issues.  So I think we can close this issue.

@zepumph anything else here? If not, please close.

**Comment by zepumph** [comment]
I think this is excellent. Thanks for the type doc.

**Comment by zepumph** [comment]
I still see a TODO for this issue in ButtonNode:

```


    // validate options - TODO: are these needed? Won't AlignBox handle this validation? https://github.com/phetsims/sun/issues/643
    assert && assert( _.includes( X_ALIGN_VALUES, options.xAlign ), 'invalid xAlign: ' + options.xAlign );
    assert && assert( _.includes( Y_ALIGN_VALUES, options.yAlign ), 'invalid yAlign: ' + options.yAlign );

```

**Comment by zepumph** [comment]
I confirmed that AlignBox covers this. This refactor wasn't possible until we centralized alignBox into ButtonNode.js
